libfirm-rs/src/nodes_gen.rs

```diff
@@ -0,0 +1,2786 @@
+            map.insert(op_code, Self::create_phi);
+            let op = bindings::get_op_Pin();
+            let op_code = bindings::get_op_code(op);
+            map.insert(op_code, Self::create_pin);
+            let op = bindings::get_op_Proj();
+            let op_code = bindings::get_op_code(op);
+            map.insert(op_code, Self::create_proj);
+            let op = bindings::get_op_Raise();
+            let op_code = bindings::get_op_code(op);
+            map.insert(op_code, Self::create_raise);
+            let op = bindings::get_op_Return();
+            let op_code = bindings::get_op_code(op);
+            map.insert(op_code, Self::create_return);
+            let op = bindings::get_op_Sel();
+            let op_code = bindings::get_op_code(op);
+            map.insert(op_code, Self::create_sel);
+            let op = bindings::get_op_Shl();
+            let op_code = bindings::get_op_code(op);
+            map.insert(op_code, Self::create_shl);
+            let op = bindings::get_op_Shr();
+            let op_code = bindings::get_op_code(op);
+            map.insert(op_code, Self::create_shr);
+            let op = bindings::get_op_Shrs();
+            let op_code = bindings::get_op_code(op);
+            map.insert(op_code, Self::create_shrs);
+            let op = bindings::get_op_Size();
+            let op_code = bindings::get_op_code(op);
+            map.insert(op_code, Self::create_size);
+            let op = bindings::get_op_Start();
+            let op_code = bindings::get_op_code(op);
+            map.insert(op_code, Self::create_start);
+            let op = bindings::get_op_Store();
+            let op_code = bindings::get_op_code(op);
+            map.insert(op_code, Self::create_store);
+            let op = bindings::get_op_Sub();
+            let op_code = bindings::get_op_code(op);
+            map.insert(op_code, Self::create_sub);
+            let op = bindings::get_op_Switch();
+            let op_code = bindings::get_op_code(op);
+            map.insert(op_code, Self::create_switch);
+            let op = bindings::get_op_Sync();
+            let op_code = bindings::get_op_code(op);
+            map.insert(op_code, Self::create_sync);
+            let op = bindings::get_op_Tuple();
+            let op_code = bindings::get_op_code(op);
+            map.insert(op_code, Self::create_tuple);
+            let op = bindings::get_op_Unknown();
+            let op_code = bindings::get_op_code(op);
+            map.insert(op_code, Self::create_unknown);
+        }
+        NodeFactory(map)
+    }
+
+    pub fn node(ir_node: *mut bindings::ir_node) -> Node {
+        Self::new().create(ir_node)
+    }
+
+    pub fn create(&self, ir_node: *mut bindings::ir_node) -> Node {
+        let op_code = unsafe { bindings::get_irn_opcode(ir_node) };
+        let f = self.0.get(&op_code).unwrap();
+        f(ir_node)
+    }
+
+    fn create_add(ir_node: *mut bindings::ir_node) -> Node {
```

Comment:
In the current libfirm_rs wrappers, each Node type (at least the ones we use) is a seperate struct, not a variant of one big Node enum. That makes it possible to define traits for "groups" of nodes, such as the ValueNode trait, so that certain functions can only accepts Nodes implementing that trait.

In this case, we would have sth like
```rust
impl NodeFactory {
   fn new_add<LHS, RHS>(lhs: LHS, rhs: RHS) -> Add
   where LHS: ValueNode, RHS: ValueNode { /*...*/}
   /*...*/
}

impl ValueNode for Add { /*...*/}
```

The advantage is that the compiler complains if you use the wrong node in the wrong place.

This of course also means that each of the `create_*` functions would need to return a specific node-struct and not an instance of `Node` enum.

I suspect that this is more difficult (impossible) to completely autogenerate, because the information of which function accepts which kind of node is only menthioned in the libfirm docs (if at all)

Reply:
I considered that. There are also structs for each node type, but there must be a way of constructing nodes in a generic way. That is what NodeFactory is for.